### ISSUE TITLE
how to share state between reducers best practice 

### ISSUE THREAD

**Issue by MaxTheMax2007** [ngrx/example-app#159]
Hello,

I went through the example application app and its a great source of learning. 

I was wondering I would need to share state between some reducers, what is the best way to do it using ngrx store ? I can't seem to find anyway to get the root reducer and the state from other reducers from within a reducer. Note that I can't use a createselector to derive data from 2 reducers because I need to mutate some information inside the reducer based on another reducer state. 

thanks a lot again

**Comment by simonpeters** [comment]
I use a service for this and merge the observables, and then have my components subscribe to that instead of directly on the store. I'm new myself so not sure if this is the right approach.

**Comment by JannicBeck** [comment]
Had exactly the same problem. I ended up passing down the dependent state to the reducer where I needed it. The reducer then looks like this: `const exampleReducer = (state, action, additionalState?) => // do stuff based on additionalState` and it is called by the reducer which has the additionalState (that may be root or further down the state tree). The pattern is also described in the link you provided.

Another option would be to pass the dependent state in the action, but I'm sending my actions over Websockets so I did not want to go this way, since I would end up with fat actions.

Last you should think about your state tree structure, if these reducers depend on each other anyways, maybe just merge them to one? Otherwise use one of the above.

**Comment by simonpeters** [comment]
@JannicBeck , the ngrx example app uses ```createSelector```  from ```reselect``` to combine 2 selectors into a new one, might be useful as wel to combine, if its applicable to your situation

https://github.com/ngrx/example-app/blob/428a4b4c28956cfc81ed60703116c388b72c079b/src/app/reducers/index.ts#L134

**Comment by JannicBeck** [comment]
@simonpeters yeah I'm also using `reselect` and its really awesome. However like @MaxTheMax2007 I need to do some checks what to do in the reducers depending on a config, which is stored elsewhere in the state tree.

**Comment by joe307bad** [comment]
Does it make sense to have a central "app state" ? Or is separating states into each reducer the best way to organize an ngrx app? What is the reasoning behind having a state for each reducer?

**Comment by benbro** [comment]
@JannicBeck how do you pass additionalState?
If I have [reducers](https://github.com/CodeSequence/ngrx-workshop/blob/step7/src/app/reducers.ts) and [search-reducer](https://github.com/CodeSequence/ngrx-workshop/blob/step7/src/app/search-reducer.ts).
How do You pass additional state to search-reducer?

**Comment by JannicBeck** [comment]
@benbro 

The example only consists of one reducer, so there is no need to pass additionalState, but If we would add a reducer which is f.e. responsible for a config, like max letters for search or other options etc. it would look like this:

```
export const root = (state: State, action: Action) => {
  return {
    search: fromSearch.reducer(state.search, action, state.config),
    config: fromConfig.reducer(state.config, action)
  };
};
```
so instead of the javascript object with reducers you just declare a root reducer, which delegates the action to the search and config reducers and providing the search reducer with the config as additional state. StoreModule then would just be: `StoreModule.provideStore(root)`

**Comment by visurel** [comment]
@JannicBeck This looks great, thanks for the response! Do you know how this can be done with the latest version of ngrx (v4)? The way to load reducers has changed to just have a map of ActionReducers and it seems like your solution doesn't work anymore. I'd love to know how this is possible. Thanks a lot! 👍

**Comment by JannicBeck** [comment]
> StoreModule.forRoot is imported once in the root module, accepting a reducer function or object map of reducer functions.

This comment in the example-app code says its possible to pass a reducer function, however you are right @visurel this does not work anymore. You could create an Issue for that, maybe they will allow passing a function again in the future, or at least update the comment.

However the example given above still works, we can just create a ActionReducerMap like this: 

`export const reducers: ActionReducerMap<State> = {
  root: rootReducer,
  routerReducer: fromRouter.routerReducer,
};`

[I created a minimal stupid example here.](https://github.com/JannicBeck/ngrx-share-state-example-app)

**Comment by edmundo096** [comment]
Reading http://redux.js.org/docs/recipes/reducers/BeyondCombineReducers.html may help as well for anyone that arrives here looking for answers on ngrx 4.
We could think that `combineReducers` is our ngrx4's `ActionReducerMap`.

**Comment by thepian** [comment]
That doesn't seem to apply as NGRX doesn't believe `getState()` to be a good idea.

**Comment by nettolicious** [comment]
If people are ending up here (as I did) search for how to do this. The current example app shows how to do this: https://github.com/ngrx/platform/tree/master/example-app

Essentially, the answer is selectors:

```
export interface UnstuffState {
  data: fromData.State;
  mainPage: fromMainPage.State;
}

export interface State extends fromRoot.State {
  unstuff: UnstuffState;
}

export const reducers: ActionReducerMap<UnstuffState> = {
  data: fromData.reducer,
  mainPage: fromMainPage.reducer,
};

export const getUnstuffState = createFeatureSelector<UnstuffState>('unstuff');

export const getDataState = createSelector(
  getUnstuffState,
  state => state.data
)

export const getMainPageState = createSelector(
  getUnstuffState,
  state => state.mainPage
)

export const getMainPageContainers = createSelector(
  getDataState,
  getMainPageState,
  (data, mainPage) => {
    return _.filter(_.values(data.containers), c => filterContainer(c, mainPage.filter));
  }
);
```

**Comment by Alex-Willenbrink** [comment]
@JannicBeck YOU BEAUTIFUL MAN! Exactly what I was looking for. I was previously handling these cases by obtaining the store state through effects, which is completely unnecessary when all I needed was state from another one of my reducers. 

Is the answer and example posted on stack overflow? I couldn't find a decent implementation of this solution anywhere.
